File: assets/csharp/run-with-it-github-update.cs
#!/usr/bin/env dotnet
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Nodes;

class Program
{
    static readonly HashSet<string> TerminalOpenOutcomes = new(StringComparer.Ordinal)
    {
        "blocked", "failed-review", "failed-merge"
    };

    static int Main(string[] args)
    {
        var parsed = ParseArgs(args);
        if (parsed == null)
        {
            PrintUsage();
            return 2;
        }

        if (parsed.Help)
        {
            PrintUsage();
            return 0;
        }

        try
        {
            if (parsed.Command == "render-comment")
            {
                Console.Write(RenderTerminalComment(parsed.ReportFile, parsed.Outcome));
                return 0;
            }

            if (parsed.Command == "update")
            {
                return UpdateGithub(parsed);
            }

            return 2;
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine(ex.Message);
            return 2;
        }
    }

    static int UpdateGithub(ParsedArgs args)
    {
        bool closeIssue = args.Outcome == "completed";
        if (args.Outcome != "completed" && !TerminalOpenOutcomes.Contains(args.Outcome))
        {
            return 0;
        }

        if (Environment.GetEnvironmentVariable("RUN_WITH_IT_GITHUB_UPDATES") == "0")
        {
            SetGithubUpdateState(args.StateFile, args.Issue, "skipped", "disabled");
            PrintStatus(args.Issue, args.Outcome, "skipped", new Dictionary<string, string?> { ["reason"] = "disabled" });
            return 0;
        }

        if (!GhExecutableExists())
        {
            SetGithubUpdateState(args.StateFile, args.Issue, "skipped", "gh-not-found");
            PrintStatus(args.Issue, args.Outcome, "skipped", new Dictionary<string, string?> { ["reason"] = "gh-not-found" });
            return 0;
        }

        if (!HasGithubRemote(args.RunRoot))
        {
            SetGithubUpdateState(args.StateFile, args.Issue, "skipped", "no-github-remote");
            PrintStatus(args.Issue, args.Outcome, "skipped", new Dictionary<string, string?> { ["reason"] = "no-github-remote" });
            return 0;
        }

        var comment = RenderTerminalComment(args.ReportFile, args.Outcome);
        var commentFile = Path.Combine(Path.GetTempPath(), $"github-comment-{Guid.NewGuid()}.md");
        File.WriteAllText(commentFile, comment);

        try
        {
            var commentResult = RunCommand("gh", new[] { "issue", "comment", args.Issue, "--body-file", commentFile }, args.RunRoot);
            if (commentResult.ExitCode != 0)
            {
                SetGithubUpdateState(args.StateFile, args.Issue, "failed", "comment-failed");
                PrintStatus(args.Issue, args.Outcome, "failed", new Dictionary<string, string?> { ["reason"] = "comment-failed" });
                return 0;
            }

            if (closeIssue)
            {
                var closeResult = RunCommand("gh", new[] { "issue", "close", args.Issue }, args.RunRoot);
                if (closeResult.ExitCode != 0)
                {
                    SetGithubUpdateState(args.StateFile, args.Issue, "failed", "close-failed");
                    PrintStatus(args.Issue, args.Outcome, "commented", new Dictionary<string, string?> { ["closed"] = "false", ["reason"] = "close-failed" });
                    return 0;
                }
            }

            string closedVal = closeIssue ? "true" : "false";
            SetGithubUpdateState(args.StateFile, args.Issue, "updated", $"commented;closed={closedVal}");
            PrintStatus(args.Issue, args.Outcome, "commented", new Dictionary<string, string?> { ["closed"] = closedVal });
            return 0;
        }
        finally
        {
            if (File.Exists(commentFile))
            {
                try { File.Delete(commentFile); } catch {}
            }
        }
    }

    static string RenderTerminalComment(string reportFile, string fallbackOutcome)
    {
        var report = LoadJson(reportFile);
        var outcome = AsString(report["outcome"]) ?? fallbackOutcome ?? "blocked";
        var summary = AsString(report["summary"]) ?? "No summary provided.";

        var verificationLines = new List<string>();
        if (report.TryGetPropertyValue("verification", out var verNode) && verNode is JsonObject verification)
        {
            var commands = new List<string>();
            if (verification.TryGetPropertyValue("commands_run", out var cmdsNode) && cmdsNode is JsonArray cmdsArray)
            {
                foreach (var cmd in cmdsArray)
                {
                    if (cmd != null) commands.Add(cmd.ToString());
                }
            }
            var evidence = AsString(verification["evidence"]) ?? "";

            string state = "unknown";
            if (verification.TryGetPropertyValue("passed", out var passedNode) && passedNode is JsonValue passedVal)
            {
                if (passedVal.GetValueKind() == JsonValueKind.True) state = "passed";
                else if (passedVal.GetValueKind() == JsonValueKind.False) state = "failed";
            }

            var commandText = commands.Count > 0 ? string.Join(", ", commands) : "unknown";
            verificationLines.Add($"State: {state}");
            verificationLines.Add($"Commands: {commandText}");
            verificationLines.Add($"Evidence: {(string.IsNullOrEmpty(evidence) ? "unknown" : evidence)}");
        }
        else
        {
            var verStr = verNode?.ToString();
            verificationLines.Add(!string.IsNullOrEmpty(verStr) ? verStr : "unknown");
        }

        var review = report["review_summary"] as JsonObject ?? new JsonObject();
        int? cycles = AsInt(review["cycles_used"]);
        var final = AsString(review["final_verdict"]) ?? "unknown";
        var reviewer = AsString(review["reviewer_model"]) ?? "unknown";

        string reviewLine;
        if (cycles == null)
        {
            reviewLine = $"Review: unknown, final verdict: {final}, reviewer model: {reviewer}";
        }
        else if (cycles.Value <= 1 && final == "approve")
        {
            reviewLine = $"Review: approve (1 cycle), final verdict: {final}, reviewer model: {reviewer}";
        }
        else
        {
            reviewLine = $"Review: revise ({cycles} cycles), final verdict: {final}, reviewer model: {reviewer}";
        }

        var tokens = report["token_usage"] as JsonObject;

        var lines = new List<string>
        {
            "## Status",
            outcome,
            "",
            "## Summary",
            summary,
            "",
            "## Verification"
        };
        lines.AddRange(verificationLines);
        lines.Add("");
        lines.Add("## Token Usage");
        lines.Add($"- Input tokens: {format_token(TokenTotal(tokens, "input"))}");
        lines.Add($"- Output tokens: {format_token(TokenTotal(tokens, "output"))}");
        lines.Add($"- Cache hit tokens: {format_token(TokenTotal(tokens, "cache"))}");
        lines.Add("");
        lines.Add("## Notes");
        lines.Add(reviewLine);

        var commitSha = AsString(report["commit_sha"]);
        if (!string.IsNullOrEmpty(commitSha))
        {
            lines.Add($"Commit: {commitSha}");
        }

        if (report.TryGetPropertyValue("merge", out var mergeNode) && mergeNode is JsonObject merge)
        {
            var mergeSha = AsString(merge["merge_sha"]);
            if (!string.IsNullOrEmpty(mergeSha))
            {
                lines.Add($"Merge: {mergeSha}");
            }
        }

        if (report.TryGetPropertyValue("blocking_reasons", out var blockNode) && blockNode is JsonArray blocking && blocking.Count > 0)
        {
            lines.Add("");
            lines.Add("## Blocking Reasons");
            foreach (var reason in blocking)
            {
                if (reason != null)
                {
                    lines.Add($"- {reason}");
                }
            }
        }

        return string.Join("\n", lines) + "\n";
    }

    static bool GhExecutableExists()
    {
        var pathEnv = Environment.GetEnvironmentVariable("PATH");
        if (string.IsNullOrEmpty(pathEnv)) return false;

        var separator = Path.PathSeparator;
        var paths = pathEnv.Split(separator, StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);
        foreach (var p in paths)
        {
            try
            {
                var fullPath = Path.Combine(p, "gh");
                if (File.Exists(fullPath))
                {
                    return true;
                }
            }
            catch {}
        }
        return false;
    }

    static bool HasGithubRemote(string runRoot)
    {
        var res = RunCommand("git", new[] { "-C", runRoot, "remote", "-v" });
        return res.ExitCode == 0 && res.Stdout.ToLowerInvariant().Contains("github.com");
    }

    static void PrintStatus(string issue, string outcome, string action, Dictionary<string, string?> fields)
    {
        var suffix = string.Join("", fields.Where(f => f.Value != null).Select(f => $"|{f.Key}={f.Value}"));
        Console.WriteLine($"STATUS|type=github-update|issue={issue}|outcome={outcome}|action={action}{suffix}");
    }

    static void SetGithubUpdateState(string stateFile, string issue, string status, string detail)
    {
        var state = LoadJson(stateFile);
        if (!state.TryGetPropertyValue("issue_registry", out var regNode) || regNode is not JsonObject registry)
        {
            registry = new JsonObject();
            state["issue_registry"] = registry;
        }
        if (!registry.TryGetPropertyValue(issue, out var entryNode) || entryNode is not JsonObject entry)
        {
            entry = new JsonObject();
            registry[issue] = entry;
        }
        entry["github_update_status"] = status;
        entry["github_update_detail"] = detail;
        entry["github_updated_at"] = DateTimeOffset.UtcNow.ToString("yyyy-MM-ddTHH:mm:ss'Z'", CultureInfo.InvariantCulture);
        SaveJson(stateFile, state);
    }

    static JsonObject LoadJson(string path)
    {
        var text = File.ReadAllText(path);
        return JsonNode.Parse(text) as JsonObject ?? new JsonObject();
    }

    static void SaveJson(string path, JsonObject content)
    {
        var parent = Path.GetDirectoryName(path);
        if (!string.IsNullOrWhiteSpace(parent))
        {
            Directory.CreateDirectory(parent);
        }

        var temp = $"{path}.tmp.{Environment.ProcessId}";
        File.WriteAllText(temp, content.ToJsonString(new JsonSerializerOptions { WriteIndented = true }) + "\n");
        File.Move(temp, path, overwrite: true);
    }

    static ParsedArgs? ParseArgs(string[] args)
    {
        if (args.Length == 0) return null;
        var parsed = new ParsedArgs();
        if (args.Contains("--help") || args.Contains("-h"))
        {
            parsed.Help = true;
            return parsed;
        }

        parsed.Command = args[0];
        if (parsed.Command != "update" && parsed.Command != "render-comment")
        {
            return null;
        }

        for (int i = 1; i < args.Length; i++)
        {
            var arg = args[i];
            if (!arg.StartsWith("--", StringComparison.Ordinal) || i + 1 >= args.Length)
            {
                return null;
            }
            var value = args[i + 1];
            i++;

            switch (arg)
            {
                case "--state-file":
                    parsed.StateFile = value;
                    break;
                case "--run-root":
                    parsed.RunRoot = value;
                    break;
                case "--issue":
                    parsed.Issue = value;
                    break;
                case "--outcome":
                    parsed.Outcome = value;
                    break;
                case "--report-file":
                    parsed.ReportFile = value;
                    break;
                default:
                    return null;
            }
        }

        if (parsed.Command == "update")
        {
            if (string.IsNullOrEmpty(parsed.StateFile) ||
                string.IsNullOrEmpty(parsed.RunRoot) ||
                string.IsNullOrEmpty(parsed.Issue) ||
                string.IsNullOrEmpty(parsed.Outcome) ||
                string.IsNullOrEmpty(parsed.ReportFile))
            {
                return null;
            }
        }
        else if (parsed.Command == "render-comment")
        {
            if (string.IsNullOrEmpty(parsed.Outcome) ||
                string.IsNullOrEmpty(parsed.ReportFile))
            {
                return null;
            }
        }

        return parsed;
    }

    static int? TokenTotal(JsonObject? tokens, string kind)
    {
        if (tokens == null) return null;
        int total = 0;
        bool found = false;
        foreach (var pair in tokens)
        {
            var keyL = pair.Key.ToLowerInvariant();
            if (kind == "input" && !keyL.Contains("input")) continue;
            if (kind == "output" && !keyL.Contains("output")) continue;
            if (kind == "cache" && !keyL.Contains("cache")) continue;

            if (pair.Value is JsonValue val && val.GetValueKind() == JsonValueKind.Number)
            {
                if (val.TryGetValue<int>(out var intVal))
                {
                    total += intVal;
                    found = true;
                }
                else if (val.TryGetValue<double>(out var dblVal))
                {
                    total += (int)dblVal;
                    found = true;
                }
            }
        }
        return found ? total : null;
    }

    static string format_token(int? value)
    {
        return value.HasValue ? value.Value.ToString(CultureInfo.InvariantCulture) : "unknown";
    }

    static string? AsString(JsonNode? node)
    {
        return node is JsonValue value && value.GetValueKind() == JsonValueKind.String
            ? value.GetValue<string>()
            : null;
    }

    static int? AsInt(JsonNode? node)
    {
        if (node is JsonValue value && value.GetValueKind() == JsonValueKind.Number)
        {
            if (value.TryGetValue<int>(out var parsed))
            {
                return parsed;
            }
            if (value.TryGetValue<double>(out var dbl))
            {
                return (int)dbl;
            }
        }
        return null;
    }

    static void PrintUsage()
    {
        Console.WriteLine("Usage: run-with-it-github-update <command> ...");
    }

    static ProcessResult RunCommand(string exe, string[] args, string? workingDir = null)
    {
        var startInfo = new ProcessStartInfo
        {
            FileName = exe,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            CreateNoWindow = true,
        };
        if (workingDir != null)
        {
            startInfo.WorkingDirectory = workingDir;
        }
        foreach (var arg in args)
        {
            startInfo.ArgumentList.Add(arg);
        }

        try
        {
            using var process = Process.Start(startInfo);
            if (process == null) return new ProcessResult { ExitCode = -1, Stdout = "", Stderr = "" };
            var stdout = process.StandardOutput.ReadToEnd();
            var stderr = process.StandardError.ReadToEnd();
            process.WaitForExit();
            return new ProcessResult { ExitCode = process.ExitCode, Stdout = stdout, Stderr = stderr };
        }
        catch (Exception ex)
        {
            return new ProcessResult { ExitCode = -1, Stdout = "", Stderr = ex.Message };
        }
    }
}

class ParsedArgs
{
    public string Command { get; set; } = string.Empty;
    public string StateFile { get; set; } = string.Empty;
    public string RunRoot { get; set; } = string.Empty;
    public string Issue { get; set; } = string.Empty;
    public string Outcome { get; set; } = string.Empty;
    public string ReportFile { get; set; } = string.Empty;
    public bool Help { get; set; }
}

class ProcessResult
{
    public int ExitCode { get; set; }
    public string Stdout { get; set; } = string.Empty;
    public string Stderr { get; set; } = string.Empty;
}

static class JsonExtensions
{
    public static JsonValueKind GetValueKind(this JsonNode? node)
    {
        return node is JsonValue value ? value.GetValueKind() : JsonValueKind.Undefined;
    }
}

File: assets/run-with-it-github-update.py
#!/usr/bin/env python3
"""Immediate GitHub issue update helper for run-with-it pool runners."""

from __future__ import annotations

import argparse
import datetime as dt
import json
import os
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path
from typing import Any


TERMINAL_OPEN_OUTCOMES = {"blocked", "failed-review", "failed-merge"}


def load_json(path: str) -> dict[str, Any]:
    with open(path, "r", encoding="utf-8") as handle:
        value = json.load(handle)
    return value if isinstance(value, dict) else {}


def save_json(path: str, value: dict[str, Any]) -> None:
    tmp_path = f"{path}.tmp.{os.getpid()}"
    with open(tmp_path, "w", encoding="utf-8") as handle:
        json.dump(value, handle, indent=2)
        handle.write("\n")
    os.replace(tmp_path, path)


def load_report(path: str) -> dict[str, Any]:
    if not os.path.exists(path) or os.path.getsize(path) == 0:
        return {}
    try:
        return load_json(path)
    except Exception:
        return {}


def set_github_update_state(state_file: str, issue: str, status: str, detail: str) -> None:
    state = load_json(state_file)
    entry = state.setdefault("issue_registry", {}).setdefault(str(issue), {})
    entry["github_update_status"] = status
    entry["github_update_detail"] = detail
    entry["github_updated_at"] = dt.datetime.now(dt.timezone.utc).replace(microsecond=0).isoformat().replace("+00:00", "Z")
    save_json(state_file, state)


def token_total(tokens: Any, kind: str) -> int | None:
    if not isinstance(tokens, dict):
        return None
    total = 0
    found = False
    for key, value in tokens.items():
        key_l = str(key).lower()
        if kind == "input" and "input" not in key_l:
            continue
        if kind == "output" and "output" not in key_l:
            continue
        if kind == "cache" and "cache" not in key_l:
            continue
        if isinstance(value, (int, float)):
            total += int(value)
            found = True
    return total if found else None


def format_token(value: int | None) -> str:
    return str(value) if value is not None else "unknown"


def render_terminal_comment(report_file: str, fallback_outcome: str) -> str:
    report = load_report(report_file)
    # Merge-recovery reports carry no token/review telemetry; fall back to the
    # sibling original issue report for those fields only (never for outcome,
    # summary, verification, or blocking reasons — those describe the recovery).
    if not report.get("token_usage") or not report.get("review_summary"):
        sibling = os.path.join(os.path.dirname(os.path.abspath(report_file)), "report.json")
        if sibling != os.path.abspath(report_file):
            original = load_report(sibling)
            if original:
                if not report.get("token_usage"):
                    report["token_usage"] = original.get("token_usage") or {}
                if not report.get("review_summary"):
                    report["review_summary"] = original.get("review_summary") or {}
                if report.get("review_skipped") is None and original.get("review_skipped") is not None:
                    report["review_skipped"] = original.get("review_skipped")
                    report["review_skip_reason"] = original.get("review_skip_reason")
    outcome = report.get("outcome") or fallback_outcome or "blocked"
    summary = report.get("summary") or "No summary provided."
    verification = report.get("verification") or {}
    if isinstance(verification, dict):
        commands = verification.get("commands_run") or []
        evidence = verification.get("evidence") or ""
        passed = verification.get("passed")
        state = "passed" if passed is True else "failed" if passed is False else "unknown"
        command_text = ", ".join(str(command) for command in commands) if commands else "unknown"
        verification_lines = [
            f"State: {state}",
            f"Commands: {command_text}",
            f"Evidence: {evidence or 'unknown'}",
        ]
    else:
        verification_lines = [str(verification) if verification else "unknown"]

    review = report.get("review_summary") or {}
    cycles = review.get("cycles_used")
    final = review.get("final_verdict") or "unknown"
    reviewer = review.get("reviewer_model") or "unknown"
    if report.get("review_skipped") is True:
        skip_reason = report.get("review_skip_reason") or "trivial-change"
        review_line = f"Review: skipped ({skip_reason})"
    elif cycles is None:
        review_line = f"Review: unknown, final verdict: {final}, reviewer model: {reviewer}"
    elif int(cycles) <= 1 and final == "approve":
        review_line = f"Review: approve (1 cycle), final verdict: {final}, reviewer model: {reviewer}"
    else:
        review_line = f"Review: revise ({cycles} cycles), final verdict: {final}, reviewer model: {reviewer}"

    tokens = report.get("token_usage") or {}
    lines = [
        "## Status",
        str(outcome),
        "",
        "## Summary",
        str(summary),
        "",
        "## Verification",
        *verification_lines,
        "",
        "## Token Usage",
        f"- Input tokens: {format_token(token_total(tokens, 'input'))}",
        f"- Output tokens: {format_token(token_total(tokens, 'output'))}",
        f"- Cache hit tokens: {format_token(token_total(tokens, 'cache'))}",
        "",
        "## Notes",
        review_line,
    ]
    if report.get("commit_sha"):
        lines.append(f"Commit: {report['commit_sha']}")
    merge = report.get("merge") or {}
    if isinstance(merge, dict) and merge.get("merge_sha"):
        lines.append(f"Merge: {merge['merge_sha']}")

    blocking = report.get("blocking_reasons") or []
    if blocking:
        lines.extend(["", "## Blocking Reasons"])
        lines.extend(f"- {reason}" for reason in blocking)

    return "\n".join(lines) + "\n"


def has_github_remote(run_root: str) -> bool:
    try:
        result = subprocess.run(
            ["git", "-C", run_root, "remote", "-v"],
            check=False,
            capture_output=True,
            text=True,
        )
    except FileNotFoundError:
        return False
    return result.returncode == 0 and "github.com" in result.stdout.lower()


def print_status(issue: str, outcome: str, action: str, **fields: Any) -> None:
    suffix = "".join(f"|{key}={value}" for key, value in fields.items() if value is not None)
    print(f"STATUS|type=github-update|issue={issue}|outcome={outcome}|action={action}{suffix}")


def update_github(args: argparse.Namespace) -> int:
    close_issue = args.outcome == "completed"
    if args.outcome != "completed" and args.outcome not in TERMINAL_OPEN_OUTCOMES:
        return 0

    if os.environ.get("RUN_WITH_IT_GITHUB_UPDATES", "1") == "0":
        set_github_update_state(args.state_file, args.issue, "skipped", "disabled")
        print_status(args.issue, args.outcome, "skipped", reason="disabled")
        return 0

    if shutil.which("gh") is None:
        set_github_update_state(args.state_file, args.issue, "skipped", "gh-not-found")
        print_status(args.issue, args.outcome, "skipped", reason="gh-not-found")
        return 0

    if not has_github_remote(args.run_root):
        set_github_update_state(args.state_file, args.issue, "skipped", "no-github-remote")
        print_status(args.issue, args.outcome, "skipped", reason="no-github-remote")
        return 0

    comment = render_terminal_comment(args.report_file, args.outcome)
    with tempfile.NamedTemporaryFile("w", encoding="utf-8", suffix=".md", delete=False) as handle:
        handle.write(comment)
        comment_file = handle.name

    try:
        comment_result = subprocess.run(
            ["gh", "issue", "comment", str(args.issue), "--body-file", comment_file],
            cwd=args.run_root,
            check=False,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )
        if comment_result.returncode != 0:
            set_github_update_state(args.state_file, args.issue, "failed", "comment-failed")
            print_status(args.issue, args.outcome, "failed", reason="comment-failed")
            return 0

        if close_issue:
            close_result = subprocess.run(
                ["gh", "issue", "close", str(args.issue)],
                cwd=args.run_root,
                check=False,
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
            )
            if close_result.returncode != 0:
                set_github_update_state(args.state_file, args.issue, "failed", "close-failed")
                print_status(args.issue, args.outcome, "commented", closed="false", reason="close-failed")
                return 0

        closed = "true" if close_issue else "false"
        set_github_update_state(args.state_file, args.issue, "updated", f"commented;closed={closed}")
        print_status(args.issue, args.outcome, "commented", closed=closed)
        return 0
    finally:
        Path(comment_file).unlink(missing_ok=True)


def render_comment(args: argparse.Namespace) -> int:
    sys.stdout.write(render_terminal_comment(args.report_file, args.outcome))
    return 0


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="run-with-it GitHub terminal update helper")
    subparsers = parser.add_subparsers(dest="command", required=True)

    update = subparsers.add_parser("update")
    update.add_argument("--state-file", required=True)
    update.add_argument("--run-root", required=True)
    update.add_argument("--issue", required=True)
    update.add_argument("--outcome", required=True)
    update.add_argument("--report-file", required=True)
    update.set_defaults(func=update_github)

    render = subparsers.add_parser("render-comment")
    render.add_argument("--outcome", required=True)
    render.add_argument("--report-file", required=True)
    render.set_defaults(func=render_comment)

    return parser


def main(argv: list[str] | None = None) -> int:
    parser = build_parser()
    args = parser.parse_args(argv)
    return args.func(args)


if __name__ == "__main__":
    sys.exit(main())
